pkg/controller/build/buildrequest/machineosbuild.go

```diff
@@ -51,15 +55,55 @@ func (m *MachineOSBuildOpts) validateForHash() error {
 		return fmt.Errorf("name mismatch, MachineConfigPool has %q, MachineOSConfig has %q", m.MachineConfigPool.Name, m.MachineOSConfig.Spec.MachineConfigPool.Name)
 	}
 
-	if m.OSImageURLConfig == nil {
-		return fmt.Errorf("misssing OSImageURLConfig")
+	return nil
+}
+
+// Validates that a MachineConfig has the necessary metadata for generating a
+// MachineOSBuild.
+func (m *MachineOSBuildOpts) validateMachineConfig() error {
+	if m.MachineConfig == nil {
+		return fmt.Errorf("missing required MachineConfig")
+	}
+
+	if !strings.HasPrefix(m.MachineConfig.Name, "rendered-") {
+		return fmt.Errorf("machineconfig %q is not a rendered MachineConfig", m.MachineConfig.Name)
+	}
+
+	requiredAnnos := []string{ctrlcommon.ReleaseImageVersionAnnotationKey, ctrlcommon.GeneratedByControllerVersionAnnotationKey}
+	for _, anno := range requiredAnnos {
+		val, ok := m.MachineConfig.Annotations[anno]
+		if !ok {
+			return fmt.Errorf("missing annotation %q on MachineConfig %q", anno, m.MachineConfig.Name)
+		}
+
+		if val == "" {
+			return fmt.Errorf("empty annotation %q value on MachineConfig %q", anno, m.MachineConfig.Name)
+		}
 	}
 
 	return nil
 }
 
 // Creates a list of objects that are consumed by the SHA256 hash.
 func (m *MachineOSBuildOpts) objectsForHash() []interface{} {
+	// Represents a private version of the OSImageURLConfig struct to keep the
+	// hashed name generation stable regardless of the input source. This means
+	// that we can eventually remove the OSImageURLConfig struct.
+	type osImageURLConfig struct {
+		BaseOSContainerImage           string
+		BaseOSExtensionsContainerImage string
+		OSImageURL                     string
+		ReleaseVersion                 string
+	}
+
+	cfg := osImageURLConfig{
+		BaseOSContainerImage:           m.MachineConfig.Spec.OSImageURL,
+		BaseOSExtensionsContainerImage: m.MachineConfig.Spec.BaseOSExtensionsContainerImage,
+		// This value is purposely left empty because the ConfigMap does not actually
+		// populate this value. However, we want the hashing to be stable.
```

Comment:
Note to reviewer: This might be a moot point since if someone is upgrading from one OCP release to another, the hashes will change. However, that means that old images may get rebuilt in the process, which is undesirable.